ARROW-3839: [Rust] Add ability to infer schema in CSV reader

nevi-me opened the pull request:
This adds the ability to infer schema on reading a CSV file.

I would have liked to also include timestamp inferrence, but looks like we don't yet have time support. I opened [ARROW-3959](https://issues.apache.org/jira/browse/ARROW-3959) to track this.

andygrove commented on rust/src/csv/reader.rs:
```diff
@@ -52,6 +53,127 @@ use crate::record_batch::RecordBatch;
+
+    // take at most 100 rows for schema inference
+    // [TODO] it's often better to read the whole file and infer its schema
+    for result in csv_reader.into_records().take(100) {
```
It would be nice to give user the ability to pass in number of records to read as an Optional where the default is to read all rows

andygrove commented on rust/src/csv/reader.rs:
```diff
@@ -52,6 +53,127 @@ use crate::record_batch::RecordBatch;
+
+        for i in 0..header_length {
+            let borrowed = record.clone();
+            my_vals[i].push(borrowed.get(i).map(|s| s.to_string()));
```
Building an in-memory list of all values seems expensive. Could we just build a set of possible types per column here instead? That way the max number of items per column is very low (limited to the number of data types)

nevi-me replied:
Done, memory usage is much better, especially for low cardinality values. I tested with a 2GB csv file, and memory usage was reasonable, although inference is slow if I din't limit max records.

andygrove replied:
I think I did a poor job explaining this... currently you build a HashSet of unique values per column (potentially pretty large) in one pass and then process that hashset to get a hashset of possible types ... what I was suggesting is why not just build the hashset of possible types as you go and then there is no need to store the actual values at all ... does that make sense?

andygrove commented on rust/src/csv/reader.rs:
```diff
@@ -149,21 +151,33 @@ fn infer_schema(mut file: File, delimiter: u8) -> Schema {
-        let record = result.unwrap();
+    match max_records {
+        Some(n) => {
+            for result in csv_reader.into_records().take(n) {
```
Instead of repeating the logic, you can just get a single iterator based on max_records like this:

```
    let records: Box<Iterator<Item = Result<StringRecord>>> = match max_records {
        Some(n) => Box::new(csv.record_iter.take(n)),
        None => Box::new(csv.record_iter)
    };

    for record  in records {

    }
```

nevi-me replied:
I had tried reusing the logic, but couldn't get it to work. 

I'm not getting your suggestion to work,

```rust
error[E0271]: type mismatch resolving `<csv_crate::StringRecordsIntoIter<std::io::BufReader<std::fs::File>> as std::iter::Iterator>::Item == std::result::Result<csv_crate::StringRecord, error::ArrowError>`
   --> src\csv\reader.rs:155:63
    |
155 |       let records: Box<Iterator<Item = Result<StringRecord>>> = match max_records {
    |  _______________________________________________________________^
156 | |         Some(n) => Box::new(csv_reader.into_records().take(n)),
157 | |         None => Box::new(csv_reader.into_records())
158 | |     };
    | |_____^ expected struct `csv_crate::Error`, found enum `error::ArrowError`
    |
    = note: expected type `std::result::Result<_, csv_crate::Error>`
               found type `std::result::Result<_, error::ArrowError>`
    = note: required for the cast to the object type `dyn std::iter::Iterator<Item=std::result::Result<csv_crate::StringRecord, error::ArrowError>>
```

andygrove replied:
OK not to worry I can make this minor optimization as a separate PR once this one is merged

paddyhoran replied:
Yes, the code in this section does seem a little repetitive...

paddyhoran commented on rust/src/csv/reader.rs:
```diff
@@ -208,7 +369,36 @@ mod tests {
+    }
+
+    #[test]
+    fn test_csv_with_schema_inferrence() {
```
nit: typo

paddyhoran commented on rust/src/csv/reader.rs:
```diff
@@ -259,7 +457,21 @@ mod tests {
+    }
+
+    #[test]
+    fn test_nulls_with_inference() {
```
It's good to check for nulls but we should test the inference logic here too.  i.e. a column with ints and floats should be inferred to be float.  We should cover each of the cases possible.

andygrove replied:
When inferring types we should not care if the values are quoted or not based on my interpretation of https://tools.ietf.org/html/rfc4180 which is as close to a specification as we have for CSV

nevi-me replied:
The RFC doesn't entirely address the concerns that I had here, and I think ignoring quoted values (if they're the only ones in otherwise non-quoted files) might lead to undesired/unintended differences. I'm mainly using Spark, Pandas and some SQL import tools that I've used to inform my gut-feel.

I'd like to do some research in the coming weeks when I'm back at work, to look at how various csv readers treat this case, and I can then update the code for that.

I propose opening a separate JIRA for this case ("Improve CSV Reader to allow specifying treatment of double quotes"), as I think it's valuable enough to spend a bit more time on ensuring that we conform with what's done in the wild. @andygrove @sunchao @paddyhoran  any opinions/disagreements?

nevi-me commented:
@andygrove @paddyhoran the test that currently (will) fail is related to the below:

Given a csv that looks like:

```csv
| c_int    | c_float | c_string | c_bool | 
|----------|---------|----------|--------| 
| 1        | 1.1     | "1.11"   | true   | 
| 2        | 2.2     | "2.22"   | true   | 
| 3        |         | "3.33"   | true   | 
| 4        | 4.4     |          | false  | 
| 5,6.6,"" | false   | ""       |        | 
```

`c_string` can either be a float or a string, but I presume that we want it to be a string in this instance. The schema inference converts it to a float, which causes the test failure.

The `csv` crate has an option on the `ReaderBuilder` to [tweak the quoting behaviour](https://docs.rs/csv/1.0.5/csv/struct.ReaderBuilder.html#method.quoting), which doesn't remove the quotes when modified. Doing that however results in an error as line 4 on the csv table above is inconsistent. 

The empty `c_string` column shouldn't be entirely empty, but should be `""`.

Should I modify it to be like line 5?

nevi-me commented:
The CI failures are parquet related, @andygrove @paddyhoran @sunchao please take another look. Thanks

sunchao commented on rust/src/csv/reader.rs:
```diff
@@ -70,21 +204,46 @@ impl Reader {
+                    batch_size,
+                }
+            }
+            None => {
```
can we remove the shared code and only do `infger_schema` when `schema` is `None`?

sunchao commented on rust/src/csv/reader.rs:
```diff
@@ -70,21 +204,46 @@ impl Reader {
     pub fn new(
         file: File,
-        schema: Arc<Schema>,
+        schema: Option<Arc<Schema>>,
```
With these many parameters, we should implement a `ReaderBuilder`.

nevi-me replied:
I implemented a `ReaderBuilder`. Please let me know if it's fine the way that I did it. I initially created builder methods that returned a `&mut ReaderBuilder` (e.g. pub fn with_delimiter(&self) -> &mut Self`, but using the methods in examples was PITA. So I used a builder that returns a new copy each time.

nevi-me commented on rust/src/csv/reader.rs:
```diff
@@ -54,6 +65,129 @@ use crate::record_batch::RecordBatch;
+}
+
+/// Infer the schema of a record
+fn infer_field_schema(string: &str) -> usize {
```
I used `arrow::datatypes::DataType` enum, which makes it clearer, I hadn't thought of doing that.

nevi-me commented on rust/src/csv/reader.rs:
```diff
@@ -54,6 +65,129 @@ use crate::record_batch::RecordBatch;
+    // return file seek back to start
+    file.seek(SeekFrom::Start(0)).unwrap();
+
+    let csv_reader = csv::ReaderBuilder::new()
```
The first times that I tried it, it wasn't working, but I got it to work now.

nevi-me commented:
Thanks @sunchao. I:

* Reverted the `Reader::new()` to what Andy had created (current master: https://github.com/apache/arrow/blob/master/rust/src/csv/reader.rs#L71)
* Created a `ReaderBuilder` that sets the various options that I've introduced (custom delimiter, max number of records for inference 
* Documented what I've added

There's still a few small issues that I picked up about our csv reader (as it stands), but I'll open separate JIRAs for them as they don't necessarily relate to schema inference.

nevi-me commented on rust/src/csv/reader.rs:
```diff
@@ -54,6 +65,129 @@ use crate::record_batch::RecordBatch;
+}
+
+/// Infer the schema of a csv file
+fn infer_schema(mut file: File, delimiter: u8, max_records: Option<usize>) -> Schema {
```
@sunchao I've renamed the method and the `max_records` to be clearer, and also added more comments on what it does.

Re the error handling and `unwrap()`, it's a bit difficult for me as I don't know what the preference with error handling is. I could return `Result<Schema>, but I'll have to write some code to convert `csv crate` errors to `ArrowError`. I'll try it out and see if I can get it right.

sunchao commented on rust/src/csv/reader.rs:
```diff
@@ -54,6 +57,120 @@ use crate::record_batch::RecordBatch;
+    // save the csv reader position after reading headers
+    let position = csv_reader.position().clone();
+
+    let my_headers: Vec<&str> = headers.iter().collect();
```
nit: can we change `my_headers` to `headers` and move it below the above `let headers = &csv_reader..`?

sunchao commented on rust/src/csv/reader.rs:
```diff
@@ -54,6 +57,120 @@ use crate::record_batch::RecordBatch;
+            2 => {
+                if possibilities.contains(&DataType::Int64)
+                    && possibilities.contains(&DataType::Float64)
+                {
```
nit: move this at the end of last line.

nevi-me replied:
`cargo fmt` says no 😜

sunchao commented on rust/src/csv/reader.rs:
```diff
@@ -78,16 +197,41 @@ impl Reader {
         }
     }
 
+    pub fn from_builder(file: File, builder: ReaderBuilder) -> Self {
```
This should be changed to a method like `build` in `ReaderBuilder`.

nevi-me replied:
Done

sunchao commented on rust/src/csv/reader.rs:
```diff
@@ -78,16 +197,41 @@ impl Reader {
     }
 
+    pub fn from_builder(file: File, builder: ReaderBuilder) -> Self {
+        // when no schema is provided, the CSV should have headers
```
How do we know that CSV should have headers when no schema is provided? should this be a pre-condition check?

nevi-me replied:
I've removed this logic, and now instead generate default column names (`column_1`, `column_2`, ...) when headers aren't provided. So we no longer need a pre-condition check.

sunchao commented on rust/src/csv/reader.rs:
```diff
@@ -200,6 +344,150 @@ impl Reader {
+    }
+
+    /// Set the CSV file's schema
+    pub fn with_schema(&self, schema: Arc<Schema>) -> ReaderBuilder {
```
Each builder method should take a `self` and return a `self`, e.g.,

```rust
pub fn with_schema(self, schema: Arc<Schema>) -> Self {
    self.schema = Some(schema)
    self
}
```

nevi-me replied:
Done

nevi-me commented on rust/src/csv/reader.rs:
```diff
@@ -54,6 +57,120 @@ use crate::record_batch::RecordBatch;
+        .from_reader(BufReader::new(file.try_clone().unwrap()));
+
+    // get header names
+    let headers = &csv_reader.headers().unwrap().clone();
```
Now using default values, this is consistent with other csv readers (like Spark and Pandas)

nevi-me commented:
@sunchao I've addressed your reviews, please take another look.

In summary, I've:

* Removed the `MAX_INFERENCE_RECORDS` and replaced with `::std::usize::MAX`
* Added error handling to address the many `unwrap()`s that I had
* Set the reader to create default column names when `has_headers` is set to false (only applies on schema inference)

andygrove commented on rust/src/csv/reader.rs:
```diff
@@ -54,6 +57,132 @@ use crate::record_batch::RecordBatch;
+
+    // get or create header names
+    // when has_headers is false, creates default column names with column_ prefix
+    let headers: Vec<String> = match has_headers {
```
nit: please use `if` rather than matching on a boolean

andygrove commented on rust/src/csv/reader.rs:
```diff
@@ -78,10 +209,9 @@ impl Reader {
         let record_iter = csv_reader.into_records();
         Reader {
-            schema: schema.clone(),
+            schema: schema,
```
nit: replace `schema: schema` with `schema`

andygrove commented on rust/src/csv/reader.rs:
```diff
@@ -54,6 +57,132 @@ use crate::record_batch::RecordBatch;
+
+        // determine data type based on possible types
+        // if there are incompatible types, use DataType::Utf8
+        match possibilities.len() {
```
can we add `bool` to the list of types that we can infer?

nevi-me replied:
It's already there. In the case where a `bool` is included with another type, we default to UTF-8.

andygrove commented:
@paddyhoran you had requested changes ... are you ok with approving this PR in light of recent commits?

nevi-me commented:
The failing test on Travis is not related to Rust. The Ruby on mac time out: `The job exceeded the maximum time limit for jobs, and has been terminated.`

kou commented:
We can ignore the Ruby CI failure because this change includes changes that affects only Rust related.

BTW, it's better that we use clean (including only the latest information) pull request description instead of keeping old information with `~~...~~` markup. Because pull request description is used as commit message in our merge tool.

nevi-me commented:
Hi @kou, I rebased, but it was the first time rebasing, so I might not have done it properly (as it looks like I still brought other changes along).

kou commented:
I've rebased.
There are many conflicts. I may choose wrong changes. Can you check again changes in this pull request?

FYI: It's better that we create a branch for each pull request instead of reusing the "master" branch.

nevi-me commented:
@kou, I've looked at the changes, made a small fix for duplicate tests. CI failures aren't Rust related

kou commented:
Merging #3325 causes conflicts again.
I'll squash your commits and rebase on the upstream master again.

kou commented:
Done.
Can you check changes again?

andygrove commented:
@nevi-me 2 tests are failing because the path to the test data changed (now under `arrow/tests` instead of `tests`)

andygrove commented on go/arrow/array/boolean.go:
```diff
@@ -45,7 +45,12 @@ func NewBooleanData(data *Data) *Boolean {
 }
 
-func (a *Boolean) Value(i int) bool { return bitutil.BitIsSet(a.values, i) }
+func (a *Boolean) Value(i int) bool {
```
@nevi-me  yes.. I think it would be best to create a new PR with just the Rust changes ... sorry for the extra work

nevi-me replied:
It's okay, coming up in a few minutes

nevi-me commented:
Closed in favour of #3349
